cmd/ctl/ca.go

```diff
@@ -310,94 +309,171 @@ Only private keys of type RSA are currently supported.
+func issueAndSaveSecret(outContext, outName, outNamespace string, upsert bool, issue func(*x509.Certificate, crypto.PrivateKey) (string, string, error)) {
+	ctx := context.TODO()
+
+	// Client for principal's kube context - it has the CA
+	caClt, err := kube.NewKubernetesClientFromConfig(ctx, globalOpts.principalNamespace, "", globalOpts.principalContext)
+	if err != nil {
+		cmdutil.Fatal("%v", err)
+	}
+
+	// Client for the kube context to write the resulting secret to - might be different
+	var outClt *kube.KubernetesClient
+	if outContext == globalOpts.principalContext || outContext == "" {
+		outClt = caClt
+	} else {
+		outClt, err = kube.NewKubernetesClientFromConfig(ctx, globalOpts.principalNamespace, "", outContext)
+		if err != nil {
+			cmdutil.Fatal("%v", err)
+		}
+	}
+
+	// Load CA keypair from a secret on the principal
+	caCert, err := tlsutil.TLSCertFromSecret(ctx, caClt.Clientset, globalOpts.principalNamespace, config.SecretNamePrincipalCA)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			cmdutil.Fatal("CA is not initialized.")
+		} else {
+			cmdutil.Fatal("Error getting CA certificate: %v", err)
+		}
+	}
+
+	// Get signer certificate
+	signerCert, err := x509.ParseCertificate(caCert.Certificate[0])
+	if err != nil {
+		cmdutil.Fatal("Could not parse CA certificate: %v", err)
+	}
 
-			_, err = clt.Clientset.CoreV1().Secrets(globalOpts.namespace).Create(ctx, secret, v1.CreateOptions{})
+	// Generate a new TLS keypair from the cert and private key
+	cert, key, err := issue(signerCert, caCert.PrivateKey)
+	if err != nil {
+		cmdutil.Fatal("Error generating certificate: %v", err)
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      outName,
+			Namespace: outNamespace,
+		},
+		Type: "kubernetes.io/tls",
+		Data: map[string][]byte{
+			"tls.crt": []byte(cert),
+			"tls.key": []byte(key),
+		},
+	}
+
+	_, err = outClt.Clientset.CoreV1().Secrets(outNamespace).Create(ctx, secret, v1.CreateOptions{})
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			cmdutil.Fatal("Could not create secret: %v", err)
+		} else if !upsert {
+			cmdutil.Fatal("Certificate exists, please use --upsert to reissue.")
+		} else {
+			_, err = outClt.Clientset.CoreV1().Secrets(outNamespace).Update(ctx, secret, v1.UpdateOptions{})
 			if err != nil {
-				if !errors.IsAlreadyExists(err) {
-					cmdutil.Fatal("Could not create secret: %v", err)
-				} else if !upsert {
-					cmdutil.Fatal("Certificate exists, please use --upsert to reissue.")
-				} else {
-					_, err = clt.Clientset.CoreV1().Secrets(globalOpts.namespace).Update(ctx, secret, v1.UpdateOptions{})
-					if err != nil {
-						cmdutil.Fatal("Could not update secret: %v", err)
-					} else {
-						fmt.Printf("Secret updated.\n")
-					}
-				}
+				cmdutil.Fatal("Could not update secret: %v", err)
 			} else {
-				fmt.Printf("Secret %s/%s created\n", globalOpts.namespace, fmt.Sprintf("%s-tls", component))
+				fmt.Printf("Secret updated.\n")
```

Comment:
Reading logs, it would be great to know namespace/name - same as for the "created" message.